Add documentation for Storage abstraction (#625)
scalar-labs/scalardb: docs/storage-abstraction.md

> # Storage abstraction

Author: Added documentation for the Storage abstraction.

> Scalar DB is middleware on top of existing storage/database systems and provides a storage/database abstraction and a storage/database-agnostic ACID transaction manager on top of the abstraction.

Reviewer: ```suggestion
Scalar DB is a middleware on top of existing storage/database systems and provides a storage/database abstraction and a storage/database-agnostic ACID transaction manager on top of the abstraction.
```

You usually use the ACID transaction manager but can also use the storage/database abstraction directly.
With the abstraction, you can write code without being aware of the storage/database implementation.

> Note that the storage abstraction operations are not transactional, and data inconsistencies can happen when failures happen during the execution of the operations.

Reviewer: ```suggestion
Note that the storage abstraction operations are not transactional, and data inconsistencies can happen when failures occur during the execution of the operations.
```
To avoid the repetition of "happen"

    if (newFromBalance < 0) {
      throw new RuntimeException(fromId + " doesn't have enough balance.");
    }

    // Update the balances
    Put fromPut =
        Put.newBuilder()
            .namespace(NAMESPACE)
            .table(TABLENAME)
            .partitionKey(Key.ofText(ID, fromId))
            .intValue(BALANCE, newFromBalance)
            .build();
    Put toPut =
        Put.newBuilder()
            .namespace(NAMESPACE)
            .table(TABLENAME)
            .partitionKey(Key.ofText(ID, toId))
            .intValue(BALANCE, newToBalance)
            .build();
    storage.put(fromPut);
    storage.put(toPut);
  }

  public int getBalance(String id) throws ExecutionException {
    // Retrieve the current balances for id
    Get get =
        Get.newBuilder()
            .namespace(NAMESPACE)
            .table(TABLENAME)
            .partitionKey(Key.ofText(ID, id))
            .build();
    Optional<Result> result = storage.get(get);

    int balance = -1;
    if (result.isPresent()) {
      balance = result.get().getInt(BALANCE);
    }
    return balance;
  }

  public void close() {
    storage.close();
  }
}
```

> Again, note that the storage abstraction operations are not transactional, and data inconsistencies can happen when failures happen during the execution of the operations.

Reviewer: ```suggestion
Again, note that the storage abstraction operations are not transactional, and data inconsistencies can happen when failures occur during the execution of the operations.
```
Ditto

TableMetadata tableMetadata =
    TableMetadata.newBuilder()
        .addColumn("c1", DataType.INT)
        .addColumn("c2", DataType.TEXT)
        .addColumn("c3", DataType.BIGINT)
        .addColumn("c4", DataType.FLOAT)
        .addColumn("c5", DataType.DOUBLE)
        .addPartitionKey("c1")
        .addClusteringKey("c2", Scan.Ordering.Order.DESC)
        .addClusteringKey("c3", Scan.Ordering.Order.ASC)
        .addSecondaryIndex("c4")
        .build();
```

Here you define columns, a partition key, a clustering key including clustering orders, and secondary indexes of a table.

Please see [Scalar DB design document - Data Model](design.md#data-model) for the details of the Scalar DB Data Model.

And then, you can create a table as follows:

```java
// Create a table "ns.tbl". It will throw an exception if the table already exists
admin.createTable("ns", "tbl", tableMetadata);

// Create a table only if it does not already exist
boolean ifNotExists = true;
admin.createTable("ns", "tbl", tableMetadata, ifNotExists);

// Create a table with options
Map<String, String> options = ...;
admin.createTable("ns", "tbl", tableMetadata, options);
```

#### Create a secondary index

> you can create a secondary index as follows:

Reviewer: ```suggestion
You can create a secondary index as follows:
```

```java
// Create a secondary index on a column "c5" of a table "ns.tbl". It will throw an exception if the secondary index already exists
admin.createIndex("ns", "tbl", "c5");

// Create a secondary index only if it does not already exist
boolean ifNotExists = true;
admin.createIndex("ns", "tbl", "c5", ifNotExists);

// Create a secondary index with options
Map<String, String> options = ...;
admin.createIndex("ns", "tbl", "c5", options);
```

#### Truncate a table

> you can truncate a table as follows:

Reviewer: ```suggestion
You can truncate a table as follows:
```

```java
// Truncate a table "ns.tbl"
admin.truncateTable("ns", "tbl");
```

#### Drop a secondary index

> you can drop a secondary index as follows:

Reviewer: ```suggestion
You can drop a secondary index as follows:
```

```java

> // Drop a secondary index on a column "c5" of a table "ns.tbl". It will throw an exception if the secondary index dose not exists

Reviewer: ```suggestion
// Drop a secondary index on a column "c5" of a table "ns.tbl". It will throw an exception if the secondary index does not exists
```

admin.dropIndex("ns", "tbl", "c5");

// Drop a secondary index only if it exists
boolean ifExists = true;
admin.dropIndex("ns", "tbl", "c5", ifExists);
```

#### Drop a table

> you can drop a table as follows:

Reviewer: ```suggestion
You can drop a table as follows:
```

```java

> // Drop a table "ns.tbl". It will throw an exception if the table dose not exists

Reviewer: ```suggestion
// Drop a table "ns.tbl". It will throw an exception if the table does not exists
```

admin.dropTable("ns", "tbl");

// Drop a table only if it exists
boolean ifExists = true;
admin.dropTable("ns", "tbl", ifExists);
```

#### Drop a namespace

> you can drop a namespace as follows:

Reviewer: ```suggestion
You can drop a namespace as follows:
```

```java

> // Drop a namespace "ns". It will throw an exception if the namespace dose not exists

Reviewer: ```suggestion
// Drop a namespace "ns". It will throw an exception if the namespace does not exists
```

admin.dropNamespace("ns");

// Drop a namespace only if it exists
boolean ifExists = true;
admin.dropNamespace("ns", ifExists);
```

#### Get a table metadata

> you can get a table metadata as follows:

Reviewer: ```suggestion
You can get a table metadata as follows:
```

Key partitionKey = Key.ofInt("c1", 10);
Key startClusteringKey = Key.of("c2", "aaa", "c3", 100L);
Key endClusteringKey = Key.of("c2", "aaa", "c3", 300L);

Scan scan =
    Scan.newBuilder()
        .namespace("ns")
        .table("tbl")
        .partitionKey(partitionKey)
        .start(startClusteringKey)
        .end(endClusteringKey)
        .projections("c1", "c2", "c3", "c4")
        .orderings(Scan.Ordering.desc("c2"), Scan.Ordering.asc("c3"))
        .limit(10)
        .build();

// Execute the Scan operation
Scanner scanner = storage.scan(scan);
```

You can omit the clustering key boundaries, or you can specify either a start boundary or an end boundary.
If you don't specify orderings, you get results ordered by clustering order you defined when creating the table.

Also, you can specify projections to choose which columns are returned, and limit to specify the number of records to return in Scan operations.

##### Handle Scanner objects

Scan operations in the storage abstraction return a `Scanner` object.

If you want to get results one by one from the Scanner object, you can use the `one()` method as follows:
```java
Optional<Result> result = scanner.one();
```

> Or, if you want to get results as `List` at a time, you can use the `all()` method as follows:

Reviewer: ```suggestion
Or, if you want to get results all at once as a `List`, you can use the `all()` method as follows:
```

```java
List<Result> results = scanner.all();
```

> Also, as Scanner implements `Iterable`, you can use `Enhanced For Loop` as follows:

Reviewer: ```suggestion
Also, as Scanner implements `Iterable`, you can use it in a for-each loop as follows:
```
"For-each" is maybe more common than the "`Enhanced For Loop`" naming.

        .limit(10)
        .build();

// Execute the Scan operation
Scanner scanner = storage.scan(scan);
```

Note that you can't specify clustering key boundaries and orderings in Scan with a secondary index.

##### Scan without a partition key to retrieve all the records of a table

You can also execute a Scan operation without specifying a partition key.

Instead of calling the `partitionKey()` method in the builder, you can call the `all()` method to scan a table without specifying a partition key as follows:

```java
// Create a Scan operation without a partition key
Key partitionKey = Key.ofInt("c1", 10);
Key startClusteringKey = Key.of("c2", "aaa", "c3", 100L);
Key endClusteringKey = Key.of("c2", "aaa", "c3", 300L);

Scan scan =
    Scan.newBuilder()
        .namespace("ns")
        .table("tbl")
        .all()
        .projections("c1", "c2", "c3", "c4")
        .limit(10)
        .build();

// Execute the Scan operation
Scanner scanner = storage.scan(scan);
```

Note that you can't specify clustering key boundaries and orderings in Scan without a partition key.

#### Put operation

`Put` is an operation to put a record specified by a primary key.

> It inserts a new record if the record doesn't exist and updates the record if the record exists.

Reviewer: ```suggestion
It behaves as an upsert operation: If it exists, it updates it; otherwise, it inserts it.
```

You need to create a Put object first, and then you can execute it with the `storage.put()` method as follows:

```java
// Create a Put operation
Key partitionKey = Key.ofInt("c1", 10);
Key clusteringKey = Key.of("c2", "aaa", "c3", 100L);

Put put =
    Put.newBuilder()
        .namespace("ns")
        .table("tbl")
        .partitionKey(partitionKey)
        .clusteringKey(clusteringKey)
        .floatValue("c4", 1.23F)
        .doubleValue("c5", 4.56)
        .build();

// Execute the Put operation
storage.put(put);
```

> You also put a record with null values as follows:

Reviewer: ```suggestion
You can also put a record with null values as follows:
```
